src/Development/IDE/Core/Completions.hs

```diff
@@ -0,0 +1,565 @@
+      rdrElts = globalRdrEnvElts rdrEnv
+
+      getCompls :: [GlobalRdrElt] -> ([CompItem],QualCompls)
+      getCompls = foldMap getComplsForOne
+
+      getComplsForOne :: GlobalRdrElt -> ([CompItem],QualCompls)
+      getComplsForOne (GRE n _ True _) =
+        case lookupTypeEnv typeEnv n of
+          Just tt -> case safeTyThingId tt of
+            Just var -> ([varToCompl var],mempty)
+            Nothing -> ([toCompItem curMod n],mempty)
+          Nothing -> ([toCompItem curMod n],mempty)
+      getComplsForOne (GRE n _ False prov) =
+        flip foldMap (map is_decl prov) $ \spec ->
+          let unqual
+                | is_qual spec = []
+                | otherwise = compItem
+              qual
+                | is_qual spec = Map.singleton asMod compItem
+                | otherwise = Map.fromList [(asMod,compItem),(origMod,compItem)]
+              compItem = [toCompItem (is_mod spec) n]
+              asMod = showModName (is_as spec)
+              origMod = showModName (is_mod spec)
+          in (unqual,QualCompls qual)
+
+      varToCompl :: Var -> CompItem
+      varToCompl var = CI name (showModName curMod) typ label Nothing docs
+        where
+          typ = Just $ varType var
+          name = Var.varName var
+          label = T.pack $ showGhc name
+          docs = getDocumentation tcs name
+
+      toCompItem :: ModuleName -> Name -> CompItem
+      toCompItem mn n =
+        CI n (showModName mn) Nothing (T.pack $ showGhc n) Nothing (getDocumentation tcs n)
+
+      (unquals,quals) = getCompls rdrElts
+
+  return $ CC
+    { allModNamesAsNS = allModNamesAsNS
+    , unqualCompls = unquals
+    , qualCompls = quals
+    , importableModules = moduleNames
+    }
+
+newtype WithSnippets = WithSnippets Bool
+
+toggleSnippets :: ClientCapabilities -> WithSnippets -> CompletionItem -> CompletionItem
+toggleSnippets ClientCapabilities { _textDocument } (WithSnippets with) x
+  | with && supported = x
+  | otherwise = x { _insertTextFormat = Just PlainText
+                  , _insertText       = Nothing
+                  }
+  where supported = fromMaybe False (_textDocument >>= _completion >>= _completionItem >>= _snippetSupport)
+
+-- | Returns the cached completions for the given module and position.
+getCompletions :: CachedCompletions -> TypecheckedModule -> VFS.PosPrefixInfo -> ClientCapabilities -> WithSnippets -> IO [CompletionItem]
+getCompletions CC { allModNamesAsNS, unqualCompls, qualCompls, importableModules }
+               tm prefixInfo caps withSnippets = do
+  let VFS.PosPrefixInfo { VFS.fullLine, VFS.prefixModule, VFS.prefixText } = prefixInfo
+      enteredQual = if T.null prefixModule then "" else prefixModule <> "."
+      fullPrefix  = enteredQual <> prefixText
+
+      -- default to value context if no explicit context
+      context = fromMaybe ValueContext $ getCContext pos (tm_parsed_module tm)
+
+      {- correct the position by moving 'foo :: Int -> String ->    '
```

Comment:
@serras sorry for digging up this old PR, but do you remember the rationale for this?

Reply:
Most of this code was taken from `haskell-ide-engine`, and I think that's where this originated.

Reply:
I know, I was one of the original implementors! Guess I'll have to do some more digging...